src/main/java/com/microsoft/azure/util/AzureCredentials.java

```diff
@@ -95,92 +96,186 @@ public String getTenant() {
             }
         }
 
-        public String getServiceManagementURL() {
-            if (serviceManagementURL == null) {
-                return Constants.DEFAULT_MANAGEMENT_URL;
+        String getAzureEnvironmentName() {
+            return azureEnvironmentName;
+        }
+
+        public AzureEnvironment getAzureEnvironment() {
+            if (azureEnvironment != null) {
+                return azureEnvironment;
+            }
+
+            AzureEnvironment baseEnvironment = null;
+            String envName = getAzureEnvironmentName();
+            if (Constants.ENV_AZURE.equalsIgnoreCase(envName)) {
+                baseEnvironment = AzureEnvironment.AZURE;
+            } else if (Constants.ENV_AZURE_CHINA.equalsIgnoreCase(envName)) {
+                baseEnvironment = AzureEnvironment.AZURE_CHINA;
+            } else if (Constants.ENV_AZURE_GERMANY.equalsIgnoreCase(envName)) {
+                baseEnvironment = AzureEnvironment.AZURE_GERMANY;
+            } else if (Constants.ENV_AZURE_US_GOVERNMENT.equalsIgnoreCase(envName)) {
+                baseEnvironment = AzureEnvironment.AZURE_US_GOVERNMENT;
             } else {
-                return serviceManagementURL;
+                // If the environment name is not recognized, which may happen when the user upgraded the plugin
+                // and didn't update the credentials, we try to match a known environment.
+                for (AzureEnvironment env : AzureEnvironment.knownEnvironments()) {
+                    if (matchEnvironment(env)) {
+                        baseEnvironment = env;
+                        break;
+                    }
+                }
+                if (baseEnvironment == null) {
+                    // fallback using the default Azure environment
+                    baseEnvironment = AzureEnvironment.AZURE;
+                }
             }
+
```

Comment:
How about using `readResolve` to upgrade the data?

Reply:
`readResolve` is generally used to replace the deserialized object with another object. Here `readObject` may be a better alternative approach.

I will update this.

Reply:
It turns out XStream serialization doesn't support objects with `readObject` / `writeObject` method defined.

```java
// com.thoughtworks.xstream.converters.reflection.SerializableConverter#doUnmarshal

        if (attributeName != null && !ATTRIBUTE_VALUE_CUSTOM.equals(reader.getAttribute(attributeName))) {
            throw new ConversionException("Cannot deserialize object with new readObject()/writeObject() methods");
        }
```

I will further check `readResolve`

Reply:
updated with `readResolve`